File: server/excalibur_server/src/crypto/merkle/mac.py
from io import SEEK_END, BytesIO
from typing import IO

from excalibur_server.src.crypto.exef import identify_version
from excalibur_server.src.crypto.exef.v4 import ExEFv4, HeaderV4

SUPPORTED_EXEF_VERSIONS = (3, 4)


def get_content_mac_input(exef_data: bytes | IO[bytes]) -> bytes:
    """
    Get the content MAC input for the given ExEF binary IO or raw ExEF data.

    :param exef_data: the ExEF data to get the content MAC input for
    :raises ValueError: if the ExEF version is not supported
    :return: the content MAC input for the given ExEF data
    """

    if isinstance(exef_data, bytes):
        exef_data: IO[bytes] = BytesIO(exef_data)

    exef_data.seek(0)
    version = identify_version(exef_data.read(5))

    if version not in SUPPORTED_EXEF_VERSIONS:
        raise ValueError(f"Unsupported ExEF version: {version}")

    exef_data.seek(0)
    if version == 3:
        # Just need header and footer
        header = exef_data.read(40)
        exef_data.seek(-16, SEEK_END)  # Seek 16 bytes from the back
        footer = exef_data.read(16)
        return header + footer

    # Version 4
    header = HeaderV4.from_serialized(exef_data.read(ExEFv4.header_size))
    chunk_count = header.chunk_count

    in_content = bytearray()
    for i in range(chunk_count):
        if i == chunk_count - 1:
            # Last chunk has the tag right at the end of the data
            exef_data.seek(-16, SEEK_END)
            in_content.extend(exef_data.read(16))
        else:
            # Other chunks have the tag at the end of the chunk
            offset = ExEFv4.header_size + i * (header.chunk_size + 16) + header.chunk_size
            exef_data.seek(offset)
            in_content.extend(exef_data.read(16))

    return bytes(in_content)

File: app/packages/main/src/lib/merkle/mac.ts
import { blake2b } from "@noble/hashes/blake2.js";

import { ExEFv3, ExEFv4, HeaderV4, identifyVersion } from "@lib/crypto/exef";

import { MerkleKeys } from "./keys";

/**
 * Get the content MAC input for the given ExEF data.
 *
 * @param exefData the ExEF data to get the content MAC input for
 * @throws {Error} if the ExEF version is not supported
 * @returns the content MAC input for the given ExEF data
 */
export function getContentMACInput(exefData: Buffer): Buffer {
    const version = identifyVersion(exefData);

    if (version === 3) {
        return Buffer.concat([exefData.subarray(0, ExEFv3.headerSize), exefData.subarray(-ExEFv3.footerSize)]);
    }
    if (version === 4) {
        const header = HeaderV4.fromBuffer(exefData.subarray(0, ExEFv4.headerSize));
        const chunkCount = header.chunkCount;

        const macInput = Buffer.alloc(16 * chunkCount);
        for (let i = 0; i < chunkCount; i++) {
            if (i === chunkCount - 1) {
                // Last chunk has the tag right at the end of the data
                macInput.set(exefData.subarray(-16), i * 16);
            } else {
                // Other chunks have the tag at the end of the chunk
                const offset = ExEFv4.headerSize + i * (header.chunkSize + 16) + header.chunkSize;
                macInput.set(exefData.subarray(offset, offset + 16), i * 16);
            }
        }
        return macInput;
    }

    throw new Error(`Unsupported ExEF version: ${version}`);
}

/**
 * Computes the content MAC for a given input using the provided Merkle keys.
 *
 * @param merkleKeys the Merkle keys to use for the computation
 * @param input the input to compute the MAC for
 * @returns the computed MAC
 */
export function computeContentMAC(merkleKeys: MerkleKeys, input: Buffer): Buffer {
    const hash = blake2b.create({ dkLen: 32, key: merkleKeys.content });
    hash.update(input);
    return Buffer.from(hash.digest());
}
